netket/utils/group/group.py

```diff
@@ -0,0 +1,266 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Ignore false-positives for redefined `product` functions:
+# pylint: disable=function-redefined
+
+import numpy as np
+from dataclasses import dataclass
+
+from .semigroup import SemiGroup, Element, Identity
+from netket.utils import struct, HashableArray
+from netket.utils.float import comparable, prune_zeros
+from netket.utils.types import Array, DType, Shape
+from typing import Tuple, List
+
+
+@struct.dataclass
+class Group(SemiGroup):
```

Comment:
I think that, except for very few things, this class assumes that the group is finite.
So we better call this `FiniteGroup<:FiniteSemiGroup` ?

Reply:
Is there any possible use case in NetKet for a non-finite group?

Reply:
yes that's what I was discussing yesterday about with @PhilipVinc, I would like to make sure that we have an implementation that supports groups that can be potentially very large or even infinite (say, the permutation group)

Reply:
so I think the result of the discussion is that it should be possible to implement things as lazy multiplications while keeping the current base class?

Reply:
I feel that this APIs problems have already been addressed https://emlp.readthedocs.io/en/latest/notebooks/3new_groups.html#Accelerating-the-solver-using-lazy-matrices

Reply:
We can rename this to `FiniteGroup<:FiniteSemiGroup` as @PhilipVinc proposed it.

I do not really see a performance problem here for larger groups. For instance, I can work with the full space group of a 100x100 grid on my desktop computer without issues. (At that point, doing VMC and even just storing all the spin samples becomes a problem, but that is independent of the group code here.)

The memory footprint could be reduced for some groups if they can be specified in terms of a finite set of generators. That is indeed not something supported by the current group classes and would need to be implemented separately. (Which I think is not a limitation but a natural consequence of different requirement for different use cases.)

Reply:
I'm happy with renaming it to `Finite*`, but I definitely won't try to implement anything beyond what's already there, since there isn't an actual use case for anything else yet.

> That is indeed not something supported by the current group classes [EDIT: Although the Group and SemiGroup interfaces are general enough to support this,

The main thing about `Group` is adding functionality for stuff like product tables and representation theory – that wouldn't work without explicitly listing group elements. However, as @femtobit points out the memory footprint of these symmetry groups is negligible compared to the data we handle.

> so I think the result of the discussion is that it should be possible to implement things as lazy multiplications while keeping the current base class?

You probably can do some things (the space group irrep stuff is an example, but there we assume a very specific group structure), but to do more complex things for a generic group, you probably need to list it explicitly.

> I feel that this APIs problems have already been addressed https://emlp.readthedocs.io/en/latest/notebooks/3new_groups.html#Accelerating-the-solver-using-lazy-matrices

I'm not sure what this library would do for us. It certainly doesn't help with producing representations for permutation groups, which is what we mostly need for geometrical symmetries. It might generate the regular representation automatically (but not even that is very clear): treating that as if it was an irrep is a computational nightmare (it leads to massive memory wastes in our current GCNN, which we're thinking to fix with @chrisrothUT). Doing any better requires one to provide irrep matrices oneself, which is **hard** (but I've just implemented it).

I think this library is aimed to deal with Lie groups (where their paradigm of tensor-producting a single intelligible irrep is sensible); stuff like alternating or symmetry groups also kinda fit in that paradigm, but not "sparse" permutation groups like a space group.